Cmake build

pelahi opened the pull request:
This adds a cmake build to the code, along with a yaml based compilation configuration file. In this way, we can autodiscover necessary packages and it makes it easier to include CUDA/HIP and add relevant logic in. It also removes the need to combine perl, python, and automake to build the code. 

It is currently still WIP. There are currently issues in compiling the code unrelated to the addition of cmake but rather in the code base itself. For example, there are some issues with use of `#define` that should be changed in `CELib.h`:
```c
// this is not a good idea since it will often default to a float
#define ENX (DBX * STRETCHX * EN)
#define ENY (DBY * STRETCHY * EN)
#define ENZ (DBZ * STRETCHZ * EN)

// better to type cast, though even here I think I should add ceil
static const int ENX = (DBX * STRETCHX * EN);
static const int ENY = (DBY * STRETCHY * EN);
static const int ENZ = (DBZ * STRETCHZ * EN);
```

More critically, need to look into how compilation flags are interrelated and which ones just don't function. An example of a possible compilation statement that doesn't work at all is 

There are also compilation errors even if this compilation flag is turned off, which might be due to other flags needing to be turned on. 
```
/Users/ela017/myresearch/codes/arepo_solas/src/mesh/criterion_derefinement.c:186:44: warning: use of GNU address-of-label extension [-Wgnu-label-as-value]
  186 |   if(jeans_number > 1.5 * JEANS_REFINEMENT && P[i].Mass < 0.5 * All.TargetGasMass)
      |                                            ^
/Users/ela017/myresearch/codes/arepo_solas/src/mesh/criterion_derefinement.c:186:48: error: expression cannot be followed by a postfix '[' operator; add parentheses
  186 |   if(jeans_number > 1.5 * JEANS_REFINEMENT && P[i].Mass < 0.5 * All.TargetGasMass)
      |                                            ~~~~^
      |                                            (   )
/Users/ela017/myresearch/codes/arepo_solas/src/mesh/criterion_derefinement.c:186:48: warning: subscript of a pointer to void is a GNU extension [-Wgnu-pointer-arith]
  186 |   if(jeans_number > 1.5 * JEANS_REFINEMENT && P[i].Mass < 0.5 * All.TargetGasMass)
      |                                            ~~~~^
/Users/ela017/myresearch/codes/arepo_solas/src/mesh/criterion_derefinement.c:186:51: error: member reference base type 'void' is not a structure or union
  186 |   if(jeans_number > 1.5 * JEANS_REFINEMENT && P[i].Mass < 0.5 * All.TargetGasMass)
      |                                            ~~~~~~~^~~~~
/Users/ela017/myresearch/codes/arepo_solas/src/mesh/criterion_derefinement.c:186:47: error: use of undeclared label 'P'
  186 |   if(jeans_number > 1.5 * JEANS_REFINEMENT && P[i].Mass < 0.5 * All.TargetGasMass)
```

Currently, I'm just using an incomplete `template_config.yaml` but really should be able to enforce proper interdependency of flags via cmake.

pelahi committed:
Preliminary Cmake build system added

- have added relevant cmake files and associated tools
- need to update the template_config.yaml file to have all possible options present
- also need to address the issue of making relevant compilation_options*.c files and building those

pelahi committed:
Updates to cmake build

- updated how yaml is parsed for configuration of build
- still need updates to how some configurations are handled like grackle
- still need to add the compilation info source code to the build

pelahi committed:
Bug fixes and adding in celib

- Adding in the celib build
- still have issues compiling and there appears to be some code updates required for proper compilation.

pelahi commented:
@doctorcbpower , I still need to add the `compilation_info` autogenerated files to the compilation as well.

pelahi commented:
If you have any comments on this approach let me know.

s-balu commented:
Hi @pelahi  @doctorcbpower  just jumping in here. Regarding the `CELib`, the plan is to move away from those and move to our own tabulated values. @mnickolas1 is looking into this.

The primary  motivation for this is that we want to do continuous wind injections as opposed to only at ~5 Myr using `CELib`

pelahi committed:
More cmake oriented updates along with template config

pelahi committed:
Bug fixes to code

pelahi committed:
Fixed execution parsing yaml config

pelahi committed:
Updated memory associated with ewald correction

Before there were several define statements setting the size of the
ewald table along with a define for the arrays associated with the
correction. This mean that the Ewald table was always on the stack.

The are issues with this:
- newer compilers are likely to complain about multiplications for the
  extent of the ewald arrays being a float and hence not compiling
- stack size could be quite large for large tables
- complaints about compilers not knowing the size of an array at compile time.
- also using 3d array assuming compiler will make a nice contiguous
  block (should happen with optimisation) but this is unnecessary. Just
  make a 1 D array addressed appropriately.

So to fix this issue, I have made static const variables and dynamically
allocated 1 D arrays on the heap. This means indexing has changed (need
to check no errors introduced) and need to clean-up after a run (need to
check this is called at the correct point).

This update means could can compile on Setonix (gcc-13) and other
machines where I have tested gcc-14, gcc-15.

Should test newer clang as well.

Also need to check force calculations are still correct.

s-balu commented:
Hi @pelahi ,

I have been trying to compile this branch on NCI Gadi. I managed to get this working with `intel-compilers` on Gadi. My workflow:
1. `mkdir build && cd build`
2.  `module purge; module load intel-compiler intel-mpi intel-mkl fftw3 gsl intel-tbb parmetis metis hdf5/1.10.7p`
3.  `cmake -DGMP_ROOT=$HOME/gmp ..`; I had to compile GMP myself as I couldn't find it on Gadi as a module
4. `make -j4`

But I tried the same workflow above with gcc and openmpi, and it was not working:
1. ` module purge; module load cmake gcc openmpi fftw3 gsl parmetis metis hdf5/1.10.7p` with ` cmake -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx -DGMP_ROOT=$HOME/gmp ..` runs fine but the `make -j4` compiles only upto CeLIB and then crashes.
2. I tried on the `ewald-correction-mem-update` (after going through the error messages from above) and this does start compiling the arepo source code but still crashes. Screenshot attached:
<img width="826" height="183" alt="Screenshot 2025-10-17 at 8 34 02 AM" src="https://github.com/user-attachments/assets/826e814e-ecf6-476d-b32b-73fbcacd532f" />

pelahi commented:
Can you `make VERBOSE=1` and also look at the linking flags?

s-balu committed:
Fix: add missing math library (-lm) to GSL linkage

Linker failed with “undefined reference to symbol 'fmin@@GLIBC_2.2.5'”
because the math library (libm) was not explicitly linked when using
GSL. Added `m` to the link_libraries call in the arepo_gsl macro to
ensure standard math functions (e.g. fmin, fmax, sqrt) are correctly
resolved during build.

s-balu commented:
Hi @pelahi thanks this helped figure out the issue. It was a missing proper linkage to math library as part of GSL. I have pushed a commit to the `ewald_branch` that I can confirm works for both the gcc and intel-compilers.

s-balu committed:
cmake: make AREPO_YAML_CONFIG user-overridable

- Added AREPO_YAML_CONFIG as a cached FILEPATH variable
- Default set to ${CMAKE_CURRENT_SOURCE_DIR}/template_config.yaml
- Allows overriding at configure time via:
  cmake -DAREPO_YAML_CONFIG=/path/to/config.yaml

pelahi commented:
Hi @s-balu , does the commit relate to cmake? Or is it pure source code?

s-balu commented:
Hi @pelahi  this is a cmake fix. But I was pushing to the [ewald-correction-mem-update](https://github.com/solas-sims/arepo_solas/tree/ewald-correction-mem-update) branch and not to the `cmake-build`.

I am attaching the error message when building the `make-build` branch with ` module purge; module load cmake gcc openmpi fftw3 gsl parmetis metis hdf5/1.10.7p` modules. This is why I switched to the `ewald-correction-mem-update` branch.

s-balu commented:
1. Do you suggest to stick to the `cmake-build` branch? I see from your commit message that we need to check your ewald-implementation is accurate. 
2. Also are there any specific compilers that you recommend? I have tried gcc and intel-compilers? Do you also suggest clang?

pelahi commented:
So for the https://github.com/solas-sims/arepo_solas/commit/08641668035d0c0504596ea759b24cbfb7eb574b commit, I don't understand the need to add `-lm` for gsl but I will look into this. 
And the other commit is a minor change so I can just add this to the branch as a cherry picked commit or just copy the change.

s-balu commented:
Hi @pelahi  I am running into issues while compiling with the `HAVE_HDF5` flag. I can confirm that this happens for both `gcc` as well as `intel-compilers`. Commenting here as I think this is related to the cmake changes and not the source code. Did you notice this issue?

Attached is a screenshot of the error. Apparently the HDF5 API has changed. I can confirm that if I make changes to `H5Eset_auto(H5E_DEFAULT, NULL, NULL);` for instance [here](https://github.com/solas-sims/arepo_solas/blob/33859fbf4d6d97e8da4243ad80ca5b59baf80efa/src/init/begrun.c#L115) it gets fixed. But then we enter the `io.c` and there are a ton of places where this error is triggered. Any suggestions?

My loaded modules are:
` module purge; module load intel-compiler intel-mpi intel-mkl fftw3 gsl intel-tbb parmetis metis hdf5/1.10.7p`
<img width="1087" height="89" alt="Screenshot 2025-10-21 at 3 26 39 PM" src="https://github.com/user-attachments/assets/1eaed1b9-5a6c-4059-af41-1f1131f29afe" />

pelahi commented:
This isn't to do with cmake but HDF5 API versions that are allowed. It isn't clear what HDF5 API the code needs. Does it only work with HDF5 parallel with API <= 1.12? I don't know. I can add version requirements for the HDF5 package though.

pelahi commented:
For instance, on Setonix I have no compilation issues with HDF5 1.14 + parallel built with the 1.12 API.

pelahi committed:
Update to cmake HDF5 search

pelahi commented:
@s-balu I have updated the ewald branch with a revised HDF5 cmake search.

s-balu commented:
Hi @pelahi thanks for looking into this, but unfortunately I am still having the same issue as before.
Looking at the Makefile, it seems like the API needs to be 1.6?
https://github.com/solas-sims/arepo_solas/blob/680882876ae375b1314736cdb7133c7d7b24a02e/Makefile#L76

pelahi commented:
This is interesting as my compilation is working with the 1.12 API ... Can you `make VERBOSE=1` and post what the compilation is? And can you check what `h5cc -showconfig` reports ?

s-balu commented:
Hi @pelahi sorry about the delay in getting back about this. After playing around with this, I feel one issue is that the Arepo cannot handle parallel hdf5? See [here](https://github.com/solas-sims/arepo_solas/blob/680882876ae375b1314736cdb7133c7d7b24a02e/Makefile#L100)

From the `make VERBOSE=1` with the parallel hdf5 loaded, it triggers a bunch of the following errors:
<img width="1210" height="282" alt="Screenshot 2025-10-28 at 10 22 04 AM" src="https://github.com/user-attachments/assets/d765c440-bfc2-4865-8b50-4cb2a49eb030" />

I cannot access `h5cc` on Gadi when the parallel hdf5 is loaded; only when the serial one is loaded.

s-balu commented:
So I replaced the `HDF5` section in the `cmake/AREPOCMakeUtils.cmake` with:

```
macro(arepo_find_hdf5)
    set(HDF5_PREFER_PARALLEL FALSE)
    if (HDF5_ROOT)
        set(ENV{HDF5_ROOT} ${HDF5_ROOT})
    endif()
    message("HDF5 enabled, finding serial HDF5 (1.6 API)...")
    find_package(HDF5 1.10 COMPONENTS C REQUIRED)
    if (HDF5_FOUND)
        include_directories(${HDF5_INCLUDE_DIRS})
        link_libraries(${HDF5_LIBRARIES})
        set(AREPO_HAS_HDF5 Yes)
        add_definitions(-DH5_USE_16_API)
    else()
        message(FATAL_ERROR "HDF5 not found")
    endif()
endmacro()
```

which almost works but is failing with:

<img width="812" height="245" alt="Screenshot 2025-10-28 at 10 32 34 AM" src="https://github.com/user-attachments/assets/56f7ffd2-ace3-4ff8-9bd7-be1d0bdc4491" />

pelahi commented:
Hi @s-balu , so I wouldn't use the makefile for guidance as it seems very haphazardly put together. Plus, as I mentioned, I am able to compile with parallel hdf, and hdf5 1.14.x but with the 1.12 API. Perhaps we are building different builds. 
I note that the error you are getting now has nothing to do with HDF5 but for some reason the `gitversion` and I realise I am missing the processing of the `git_version.sh` script. I'll need to update the cmake build to handle what is done by this bash script.

s-balu commented:
@pelahi  this by no chance shouldn't be something that is specific to how hdf5 is setup on Gadi would it?

pelahi commented:
It certainly could be something in how gadi setup HDF5. For example for Setonix, arepo is compiling with hdf5 parallel 
```
h5cc -showconfig
        SUMMARY OF THE HDF5 CONFIGURATION
        =================================

General Information:
-------------------
                   HDF5 Version: 1.14.5
                  Configured on: 2025-07-11
                  Configured by: Unix Makefiles
                    Host system: Linux-6.4.0-150600.23.53-default
              Uname information: Linux
                       Byte sex: little-endian
             Installation point: /software/setonix/2025.08/software/linux-sles15-zen3/gcc-14.2.0/hdf5-1.14.5-vzox6p773gcih6t6hvhra6lzvyv5wimy

Compiling Options:
------------------
                     Build Mode: Release
              Debugging Symbols: OFF
                        Asserts: OFF
                      Profiling: OFF
             Optimization Level: OFF

Linking Options:
----------------
                      Libraries:
  Statically Linked Executables: OFF
                        LDFLAGS:
                     H5_LDFLAGS:
                     AM_LDFLAGS:
                Extra libraries: m;dl;/opt/cray/pe/mpich/8.1.32/ofi/gnu/12.3/lib/libmpi_gnu_123.so;/opt/xpmem/lib64/libxpmem.so
                       Archiver: /usr/bin/ar
                       AR_FLAGS:
                         Ranlib: /usr/bin/ranlib

Languages:
----------
                              C: YES
                     C Compiler: /software/setonix/2025.08/spack/lib/spack/env/gcc/gcc 14.2.0
                       CPPFLAGS:
                    H5_CPPFLAGS:
                    AM_CPPFLAGS:
                         CFLAGS:   -std=c99 -Wno-error=implicit-function-declaration -fstdarg-opt -fdiagnostics-urls=never -fno-diagnostics-color -fmessage-length=0
                      H5_CFLAGS: -Wall;-Wcast-qual;-Wconversion;-Wextra;-Wfloat-equal;-Wformat=2;-Winit-self;-Winvalid-pch;-Wmissing-include-dirs;-Wshadow;-Wundef;-Wwrite-strings;-pedantic;-Wno-c++-compat;-Wbad-function-cast;-Wcast-align;-Wformat;-Wimplicit-function-declaration;-Wint-to-pointer-cast;-Wmissing-declarations;-Wmissing-prototypes;-Wnested-externs;-Wold-style-definition;-Wpacked;-Wpointer-sign;-Wpointer-to-int-cast;-Wredundant-decls;-Wstrict-prototypes;-Wswitch;-Wunused-but-set-variable;-Wunused-variable;-Wunused-function;-Wunused-parameter;-finline-functions;-Wno-aggregate-return;-Wno-inline;-Wno-missing-format-attribute;-Wno-missing-noreturn;-Wno-overlength-strings;-Wlarger-than=2560;-Wlogical-op;-Wframe-larger-than=16384;-Wpacked-bitfield-compat;-Wsync-nand;-Wno-unsuffixed-float-constants;-Wdouble-promotion;-Wtrampolines;-Wstack-usage=8192;-Wmaybe-uninitialized;-Wno-jump-misses-init;-Wstrict-overflow=2;-Wno-suggest-attribute=const;-Wno-suggest-attribute=noreturn;-Wno-suggest-attribute=pure;-Wno-suggest-attribute=format;-Wdate-time;-Warray-bounds=2;-Wc99-c11-compat;-Wincompatible-pointer-types;-Wint-conversion;-Wshadow;-Wduplicated-cond;-Whsa;-Wnormalized;-Wnull-dereference;-Wunused-const-variable;-Walloca;-Walloc-zero;-Wduplicated-branches;-Wformat-overflow=2;-Wformat-truncation=1;-Wrestrict;-Wattribute-alias;-Wshift-overflow=2;-Wcast-function-type;-Wmaybe-uninitialized;-Wcast-align=strict;-Wno-suggest-attribute=cold;-Wno-suggest-attribute=malloc;-Wattribute-alias=2;-Wmissing-profile;-Wc11-c2x-compat
                      AM_CFLAGS:
               Shared C Library: YES
               Static C Library: YES

                        Fortran: ON
               Fortran Compiler: /software/setonix/2025.08/spack/lib/spack/env/gcc/gfortran 14.2.0
                  Fortran Flags:
               H5 Fortran Flags:
               AM Fortran Flags:
         Shared Fortran Library: YES
         Static Fortran Library: YES
               Module Directory: /tmp/pawsey0001/spack/setonix/2025.08/software/spack/build_stage/spack-stage-hdf5-1.14.5-vzox6p773gcih6t6hvhra6lzvyv5wimy/spack-build-vzox6p7/mod

                            C++: OFF
                   C++ Compiler:
                      C++ Flags:
                   H5 C++ Flags:
                   AM C++ Flags:
             Shared C++ Library: YES
             Static C++ Library: YES

                           JAVA: OFF
                  JAVA Compiler:

Features:
---------
                     Parallel HDF5: ON
  Parallel Filtered Dataset Writes: ON
                Large Parallel I/O: ON
                High-level library: ON
Dimension scales w/ new references:
                  Build HDF5 Tests: OFF
                  Build HDF5 Tools: ON
                   Build GIF Tools: OFF
                      Threadsafety: ON
               Default API mapping: v112
    With deprecated public symbols: ON
            I/O filters (external):  DEFLATE
                  _Float16 support: ON
                     Map (H5M) API: OFF
                        Direct VFD: OFF
                        Mirror VFD: OFF
                     Subfiling VFD: OFF
                (Read-Only) S3 VFD: OFF
              (Read-Only) HDFS VFD: OFF
    Packages w/ extra debug output:
                       API Tracing: OFF
              Using memory checker: OFF
                  Use file locking: best-effort
         Strict File Format Checks: OFF
      Optimization Instrumentation:
```

pelahi commented:
You'll see that this HDF5 library is parallel, no c++ but has the 1.12 API.

pelahi commented:
It also works with an HDF5 library build to have the 1.8 API (`h5cc -showconfig | grep "Default API"` gives `Default API mapping: v18`). 
Now if I load a different build of HDF5, one without parallel, and alter the cmake so that it doesn't require it, I can still compile.

pelahi commented:
also, if you only have `h5cc` try `h5pcc` when parallel is loaded. It is the same tool (but for parallel, though it is weird that the build didn't build the serial compilation as well.

pelahi commented:
Hi @s-balu , so looking at the code, looks like you are correct, there is no parallel HDF5. This is a big issue we should address before trying to run very large simulations. Thoughts?

pelahi committed:
updates to cmake related to HDF5 along with some related to profile util and gpu functionality

pelahi committed:
Update to code to add profile util, with goal of profiling cpu and also gpu when running

pelahi committed:
Fix, added incorrect profile util directory

pelahi committed:
Updated profile util git submodule

pelahi committed:
Added cmake functionality to store git commit info

s-balu commented:
Hi @pelahi  thanks for looking into this. I will do test once I am in the office and let you know. And about the parallel HDF5, Arepo/Gadget, as you know, is able to split a snapshot into multiple files and these writes happen concurrently. So maybe not having the parallel hdf5 was a conscious design choice. Do you agree @doctorcbpower ?

pelahi committed:
cmake fix for fftw3

pelahi committed:
Fix how hdf5 compile info is produced

pelahi committed:
Merge pull request #16 from solas-sims/ewald-correction-mem-update

Ewald correction mem update

pelahi commented:
Hi @s-balu @doctorcbpower are you happy to merge this into the main branch? It now has a fix to speed up the ewald calculation, a cmake build along with the older build system.

s-balu committed:
Fix: HDF5 API, executable name

Arepo expects the (older) 1.6 API for the HDF5. Also renamed the
executable to Arepo so that it is compatible with older build method.

s-balu commented:
Hi @pelahi  just a small patch for the HDF5 API and renaming the executable to Arepo instead of arepo (mostly so that my submission scripts works for both build methods).

@doctorcbpower this branch has been tested against both gnu and intel family of modules/compilers on Gadi. I am happy for this to be merged to main. note that we also have the older `Makefile` method working alongside as well. 

I will need to update the grackle integration to take advantage of the cmake build but that is a separate issue.

s-balu commented:
Also, @pelahi  do we need to enforce the C11 standard for the cmake build before merging?

pelahi committed:
Update template config example

pelahi committed:
Update to ensure fftw3 use in cmake

pelahi committed:
Add c11 standard

pelahi committed:
Minor fix to ensure defines are not used but rather static const variables

pelahi committed:
show an example of cmake configuration

pelahi commented:
Ah, yes I forgot about that. Added. Now also have a few other fixes.

s-balu commented:
Merging this branch. I will update the wiki with info about compiling using cmake sometime this week.
